File: FUWEB/PHP/app/Http/Controllers/playerController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Hash;
use Illuminate\Support\Facades\Log;
use App\Http\Controllers\Controller;
use Illuminate\Support\Str;
use Firebase\JWT\JWT;
use Firebase\JWT\Key;

class playerController extends Controller
{
    /**
     * Get a player by their ID.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function getPlayerById(Request $request)
    {
        $userId = $request->route('id');

        $inputData = ['userId' => (int)$userId];

        try {
            $dbOutput = DB::select('CALL sp_getPlayer(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if(isset($result['data'])){
             $data = $result['data'];

                return response()->json([
                    'message' => $message,
                    'data' => $data,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }

    /**
     * Get all players.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function getAllPlayers(Request $request)
    {
        try {
            $dbOutput = DB::select('CALL sp_getPlayers()');
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];
            $data = $result['data'];

            return response()->json([
                'message' => $message,
                'data' => $data,
            ], $statusCode);
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }

    /**
     * Register a new player.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function register(Request $request)
    {
        try {
            $username = $request->input('username');
            $inputPassword = $request->input('input_password');

            if (empty($username) || empty($inputPassword)) {
                return response()->json(['error' => 'Invalid input, object invalid'], 400);
            }

            $salt = bin2hex(random_bytes(32));

            $passwordBin = hash_pbkdf2("sha512", $inputPassword, $salt, 10000, 512, true);
            $password = bin2hex($passwordBin);
            $inputData = ['username' => $username, 'password' => $password, 'salt' => $salt];
            $dbOutput = DB::select('CALL sp_register(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if (isset($result['user'], $result['new_world_created'], $result['new_world_id'])) {

                $user = $result['user'];
                $newWorldCreated = $result['new_world_created'];
                $newWorldId = $result['new_world_id'];
                if ($newWorldCreated === 1) {
                    $this->generateWorld($newWorldId);
                }

                return response()->json([
                    'message' => $message,
                    'user' => $user,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }
    
    /**
     * Login a player.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function login(Request $request)
    {
        try {
            $username = $request->input('username');
            $inputPassword = $request->input('input_password');

            $saltOutput = DB::select('CALL sp_getSalt(?)', [$username]);
            $salt = $saltOutput[0]->salt;

            if (empty($username) || empty($inputPassword) || $salt === null) {
                return response()->json(['error' => 'Invalid input, object invalid'], 400);
            }

            $password = $this->validatePassword($inputPassword, $salt);
            $token = $this->generateAuthToken($username);
            $inputData = ['username' => $username, 'password' => $password, 'token' => $token];
            $dbOutput = DB::select('CALL sp_login(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if (isset($result['user']) && $result['user'] !== null) {
                $user = $result['user'];
                return response()->json([
                    'message' => $message,
                    'user' => $user,
                    'token' => $token,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }
    
    /**
     * Get a player's stockpile.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function getPlayerStockpile(Request $request)
    {
        try {
            $userId = $request->input('userId');

            $inputData = ['userId' => (int)$userId];
            $dbOutput = DB::select('CALL sp_getPlayerRailways(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if (isset($result['user']) && $result['user'] !== null) {
                $data = $result['data'];
                return response()->json([
                    'message' => $message,
                    'data' => $data,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }
    /**
     * Get a player's needs.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function getPlayerNeeds(Request $request)
    {
        try {
            $userId = $request->input('userId');

            $inputData = ['userId' => (int)$userId];
            $dbOutput = DB::select('CALL sp_getPlayerNeeds(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if (isset($result['user']) && $result['user'] !== null) {
                $data = $result['data'];
                return response()->json([
                    'message' => $message,
                    'data' => $data,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }

    /**
     * Get a player's railways.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function getPlayerRailways(Request $request)
    {
        try {
            $userId = $request->input('userId');

            $inputData = ['userId' => (int)$userId];
            $dbOutput = DB::select('CALL sp_getPlayerRailways(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if (isset($result['user']) && $result['user'] !== null) {
                $data = $result['data'];
                return response()->json([
                    'message' => $message,
                    'data' => $data,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }

    /**
     * Get a player's trains.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function getPlayerTrains(Request $request)
    {
        try {
            $userId = $request->input('userId');

            $inputData = ['userId' => (int)$userId];
            $dbOutput = DB::select('CALL sp_getPlayersTrains(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if (isset($result['user']) && $result['user'] !== null) {
                $data = $result['data'];
                return response()->json([
                    'message' => $message,
                    'data' => $data,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }

    /**
     * Get a player's industries.
     *
     * @param Request $request
     * @return \Illuminate\Http\JsonResponse
     */
    public function getPlayerIndustries(Request $request)
    {
        try {
            $userId = $request->input('userId');

            $inputData = ['userId' => (int)$userId];
            $dbOutput = DB::select('CALL sp_getPlayerIndustries(?)', [json_encode($inputData)]);
            $result = json_decode($dbOutput[0]->result, true);
            $statusCode = $result['status_code'];
            $message = $result['message'];

            if (isset($result['user']) && $result['user'] !== null) {
                $data = $result['data'];
                return response()->json([
                    'message' => $message,
                    'data' => $data,
                ], $statusCode);
            } else {
                return response()->json([
                    'message' => $message,
                ], $statusCode);
            }
        } catch (\Exception $error) {
            \Log::error($error);
            return response()->json(['error' => 'Internal Server Error'], 500);
        }
    }
    

    /**
     * Validate a password using a salt.
     *
     * @param string $password
     * @param string $salt
     * @return string
     */
    private function validatePassword($password, $salt)
    {
        $key = hash_pbkdf2("sha512", $password, $salt, 10000, 512, true);
        return $hexString = bin2hex($key);
    }

    /**
     * Generate an authentication token.
     *
     * @param string $username
     * @return string
     */
    private function generateAuthToken($username)
    {
        $payload = [
            'username' => $username,
            'exp' => time() + 12 * 3600,
        ];
        $jwtKey = 'RailwayImperiumSecret';
        $token = JWT::encode($payload, $jwtKey, 'HS256');
        return $token;
    }

    /**
     * Generate a world with towns and rural businesses.
     *
     * @param int $worldId
     * @return void
     */
    private function generateWorld($worldId)
    {
        $gridSize = 1000;
        $minDistance = 50;
        $numTowns = 15;
        $numRuralBusinesses = 30;

        $towns = [];
        $ruralBusinesses = [];

        // Generate towns
        for ($i = 0; $i < $numTowns; $i++) {
            $position = $this->generateRandomPosition(array_merge($towns, $ruralBusinesses), $gridSize, $minDistance);
            $towns[] = $position;
        }

        // Generate rural businesses
        for ($i = 0; $i < $numRuralBusinesses; $i++) {
            $position = $this->generateRandomPosition(array_merge($towns, $ruralBusinesses), $gridSize, $minDistance);
            $ruralBusinesses[] = $position;
        }

        try {
            foreach ($towns as $i => $town) {
                $jsonData = json_encode([
                    'type' => 'TOWN',
                    'name' => 'Town ' . $i,
                    'position' => $town,
                    'worldId' => (int)$worldId,
                ]);
                $res = DB::select('CALL sp_createAsset(?)', [$jsonData]);
            }

            foreach ($ruralBusinesses as $i => $ruralBusiness) {
                $business = '';
        
                if ($i < 4) {
                    $business = 'RANCH';
                } elseif ($i < 9) {
                    $business = 'FIELD';
                } elseif ($i < 14) {
                    $business = 'FARM';
                } elseif ($i < 19) {
                    $business = 'LUMBERYARD';
                } elseif ($i < 24) {
                    $business = 'PLANTATION';
                } else {
                    $business = 'MINE';
                }
        
                $jsonData = json_encode([
                    'type' => 'RURALBUSINESS',
                    'name' => 'Rural Business' . $i,
                    'position' => $ruralBusiness,
                    'worldId' => (int)$worldId,
                    'business' => $business,
                ]);
                $res = DB::select('CALL sp_createAsset(?)', [$jsonData]);
            }
        } catch (\Exception $error) {
            \Log::error($error);
        }
    }

    /**
     * Generate a random position with minimum distance.
     *
     * @param array $existingPositions
     * @param int $gridSize
     * @param int $minDistance
     * @return array
     */
    private function generateRandomPosition($existingPositions, $gridSize, $minDistance)
    {
        do {
            $x = rand(0, $gridSize - 1);
            $y = rand(0, $gridSize - 1);
        } while ($this->hasMinDistance($existingPositions, $x, $y, $minDistance));

        return compact('x', 'y');
    }

    /**
     * Check if a new position has a minimum distance from existing positions.
     *
     * @param array $existingPositions
     * @param int $newX
     * @param int $newY
     * @param int $minDistance
     * @return bool
     */
    private function hasMinDistance($existingPositions, $newX, $newY, $minDistance)
    {
        foreach ($existingPositions as $position) {
            $distance = hypot($newX - $position['x'], $newY - $position['y']);
            if ($distance < $minDistance) {
                return true;
            }
        }
        return false;
    }
}

File: JAVA/api/src/main/java/railway/api/Controller/PlayerController.java
package railway.api.Controller;

import railway.api.Database.AssetDAO;
import railway.api.Database.PlayerDAO;

import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.KeySpec;
import java.util.Map;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import java.security.Key;
import java.util.Date;
import org.json.JSONArray;
import org.json.JSONObject;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.codec.Hex;
import org.springframework.security.crypto.keygen.BytesKeyGenerator;
import org.springframework.security.crypto.keygen.KeyGenerators;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class PlayerController {

    private PlayerDAO playerDAO = new PlayerDAO();

    @GetMapping("/players")
    public ResponseEntity<Object> getPlayers() {
        JSONObject result = playerDAO.getPlayers();

        int status_code = result.getInt("status_code");
        String message = result.getString("message");
        JSONArray data = result.getJSONArray("data");

        JSONObject responseJson = new JSONObject();
        responseJson.put("data", data);
        responseJson.put("message", message);

        return ResponseEntity.status(status_code).body(responseJson.toMap());
    }

    @GetMapping("/player/{id}")
    public ResponseEntity<Object> getPlayer(@PathVariable Long id) {
        JSONObject inputJSON = new JSONObject();
        inputJSON.put("userId", id);
        JSONObject result = playerDAO.getPlayerById(inputJSON);
        int status_code = result.getInt("status_code");
        String message = result.getString("message");
        JSONObject data = result.optJSONObject("data");

        JSONObject responseJson = new JSONObject();
        if (data != null) {
            responseJson.put("data", data);
        }
        responseJson.put("message", message);
        return ResponseEntity.status(status_code).body(responseJson.toMap());
    }

    @PostMapping("/register")
    public ResponseEntity<Object> register(@RequestBody Map<String, Object> requestBody) {
        String username = (String) requestBody.get("username");
        String input_password = (String) requestBody.get("input_password");
        String salt = generateRandomSalt();
        String password = hashPassword(input_password, salt);
        JSONObject inputJSON = new JSONObject();
        inputJSON.put("password", password);
        inputJSON.put("username", username);
        inputJSON.put("salt", salt);
        JSONObject result = playerDAO.register(inputJSON);
        int status_code = result.getInt("status_code");
        String message = result.getString("message");
        JSONObject user = result.optJSONObject("user");
        int new_world_created = result.optInt("new_world_created");
        int new_world_id = result.optInt("new_world_id");
        if (new_world_created == 1) {
            generateWorld(new_world_id);
        }
        JSONObject responseJson = new JSONObject();
        if (user != null) {
            responseJson.put("user", user);
        }
        responseJson.put("message", message);
        return ResponseEntity.status(status_code).body(responseJson.toMap());
    }
    @PostMapping("/login")
    public ResponseEntity<Object> login(@RequestBody Map<String, Object> requestBody) {
        String username = (String) requestBody.get("username");
        String input_password = (String) requestBody.get("input_password");
        String salt = playerDAO.getSalt(username);
        String password = hashPassword(input_password, salt);
        JSONObject inputJSON = new JSONObject();
        System.out.println(inputJSON);
        inputJSON.put("password", password);
        inputJSON.put("username", username);
        inputJSON.put("salt", salt);
        JSONObject result = playerDAO.login(inputJSON);
        int status_code = result.getInt("status_code");
        String message = result.getString("message");
        JSONObject user = result.optJSONObject("user");
        JSONObject responseJson = new JSONObject();
        if (user != null) {
            responseJson.put("user", user);
        }
        responseJson.put("message", message);
        return ResponseEntity.status(status_code).body(responseJson.toMap());
    }

    private String generateRandomSalt() {
        BytesKeyGenerator keyGenerator = KeyGenerators.secureRandom(32);
        byte[] saltBytes = keyGenerator.generateKey();
        return new String(Hex.encode(saltBytes));
    }

    private String hashPassword(String inputPassword, String salt) {
        String result = "";
        KeySpec spec = new PBEKeySpec(inputPassword.toCharArray(), Hex.decode(salt), 10000, 512);
        try {
            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512");
            byte[] hashedBytes = factory.generateSecret(spec).getEncoded();
            result = new String(Hex.encode(hashedBytes));
        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
            e.printStackTrace();
        }
        return result;
    }

    public static String generateAuthToken(String username) {
        Key SECRET_KEY = Keys.secretKeyFor(SignatureAlgorithm.HS256);
        long EXPIRATION_TIME = 12 * 3600 * 1000;
        Date expirationDate = new Date(System.currentTimeMillis() + EXPIRATION_TIME);

        String token = Jwts.builder()
                .setSubject(username)
                .setExpiration(expirationDate)
                .signWith(SECRET_KEY)
                .compact();

        return token;
    }

    private static JSONObject generateRandomPosition(JSONArray existingPositions, int gridSize, int minDistance) {
        int x, y;
        do {
            x = (int) (Math.random() * gridSize);
            y = (int) (Math.random() * gridSize);
        } while (hasMinDistance(existingPositions, x, y, minDistance));

        return new JSONObject().put("x", x).put("y", y);
    }

    // Function to check if a position has a minimum distance from existing
    // positions
    private static boolean hasMinDistance(JSONArray existingPositions, int newX, int newY, int minDistance) {
        for (Object positionObj : existingPositions) {
            JSONObject position = (JSONObject) positionObj;
            double distance = Math.hypot(newX - position.getInt("x"), newY - position.getInt("y"));
            if (distance < minDistance) {
                return true; // Too close, try again
            }
        }
        return false; // Minimum distance satisfied
    }

    // Function to generate a world with towns and rural businesses
    private static void generateWorld(int worldId) {
        AssetDAO assetDAO = new AssetDAO();
        int gridSize = 1000;
        int minDistance = 50;
        int numTowns = 15;
        int numRuralBusinesses = 30;

        JSONArray towns = new JSONArray();
        JSONArray ruralBusinesses = new JSONArray();

        // Generate towns
        for (int i = 0; i < numTowns; i++) {
            JSONObject position = generateRandomPosition(new JSONArray(towns.toString()), gridSize, minDistance);
            towns.put(position);
        }

        // Generate rural businesses
        for (int i = 0; i < numRuralBusinesses; i++) {
            JSONObject position = generateRandomPosition(new JSONArray(towns.toString()), gridSize, minDistance);
            ruralBusinesses.put(position);
        }

        for (int i = 0; i < towns.length(); i++) {
            JSONObject town = towns.getJSONObject(i);
            JSONObject jsonTown = new JSONObject()
                    .put("type", "TOWN")
                    .put("name", "Town " + i)
                    .put("position", town)
                    .put("worldId", worldId);
            assetDAO.createAsset(jsonTown);
        }

        for (int i = 0; i < ruralBusinesses.length(); i++) {
            JSONObject ruralBusiness = ruralBusinesses.getJSONObject(i);
            JSONObject jsonRuralBusiness = new JSONObject()
                    .put("type", "RURALBUSINESS")
                    .put("name", "Rural Business" + i)
                    .put("position", ruralBusiness)
                    .put("worldId", worldId);
            assetDAO.createAsset(jsonRuralBusiness);
        }
    }

}
